Image sending function

robot269006 opened the pull request:
# WHAT
Set carrierwave gem

# WHY
To allow image uploading function

robot269006 committed:
Set carrierwave

robot269006 committed:
Set image column in the table

1.Add image column to messages table
2.rake db:migrate

robot269006 committed:
Set image uploader

1.rails g uploader image
2.Enable image_uploader.rb to be used in upload folder
　-> mount_uploader :image, ImageUploader to messages.rb

robot269006 committed:
View edit (set file field)

1. Fix css structure -> buttons.scss and icons.scss removed
2. Allow file_field

robot269006 committed:
View edit

1. Fix icon display
2. Fix js dexcription accordingly

robot269006 committed:
Enable image selection

1. Set strong parameter for image
2. Added image li in _message.haml
3. Added image jbuilder
4. Uncomment, switch back on the image upload field in index
5. Set jquery function that switches on image selection when the icon
class is clicked

robot269006 committed:
Enable image upload

1. View refactor -> display:none; -> put into style sheet
2. Allow image upload

*But image not uploaded asynchronously

robot269006 commented:
![image1](https://cloud.githubusercontent.com/assets/24761206/26522879/2ac561e2-4345-11e7-9220-712f152aa667.gif)

ken-nag commented:
4点の大きな修正をお願いします
gif画像を確認したとこmessageが降順に表示されているので昇順にしましょう。
github上にpublic以下のuploadされた画像が上がってしまっているのでgitignoreに記述しましょう
非同期で投稿されたmessageの時間表示を修正しましょう
画像投稿の非同期をしましょう

robot269006 committed:
Revision (minor message display changes)

1. Set gitignore (ignore public and .env)
2. Set messages in asc order
3. Fix time display in messages

robot269006 committed:
Introduce S3 (fog and storage settings)

1. Create bucket
2. Install ‘fog’ gem
3. Change storage type from :file to :fog

robot269006 committed:
Introduce S3 (dotenv gem introduced)

1. Install ‘dotenv-rails’
2. Create .env files and write out AWS access keys there
3. Set gitignore -> ignore .env (SUPER IMPORTANT)

robot269006 committed:
Introduce S3 (Set up carrierwave)

1. Set up carrierwave.rb
2. Try uploading images by S3/carrierwave -> upload successful

robot269006 committed:
Revision (asynchro image upload)

1. Append image by buildHTML function

seoky0 commented on app/assets/javascripts/messages.js:
```diff
@@ -3,9 +3,14 @@ $(function() {
-    var date = list.append($('<li class="individualmessage__date">' + datareceiver.created_at + '<li>'));
-    var body = list.append($('<li class="individualmessage__body">' + datareceiver.body + '<li>'));
+    //date cleaning
+    var date1 = datareceiver.created_at.replace(/-/g, '/');
```
date1, 2, 3は使い捨ての変数のように感じるのですが、一つにまとめることはできないでしょうか？

seoky0 replied:
日付をJavaScriptの方で処理するのは大変そうなので、Railsの方で日付を整形してJSONに渡すのはいかがでしょうか？

robot269006 replied:
ご教示ありがとうございます。トライしてみます！

seoky0 commented on app/assets/javascripts/messages.js:
```diff
@@ -14,17 +19,20 @@ $(function() {
     e.preventDefault();
-    var textField = $('.write__text');
+    var textField = $('.group__bottom__form__text');
     var message = textField.val();
```
formDataを使うのであれば不要ですね

robot269006 replied:
確かにフォームデータ全て送れますね。削除します。

robot269006 committed:
Revision (Date description)

1. Date description revised
2. Delete retrieving textfield, since formData is used

seoky0 commented on .gitignore:
```diff
@@ -11,3 +11,5 @@
 /log/*
 !/log/.keep
 /tmp
+/public
```
これだとpublic以下でもgitで管理したいものがあるときに修正することになります。
本当に管理したくないものだけを指定しましょう。

robot269006 replied:
承知しました、carrierwave以前は直接/public以下に保管してたのでそのままになってました。削除します。

seoky0 commented on app/controllers/messages_controller.rb:
```diff
@@ -4,7 +4,7 @@ class MessagesController < ApplicationController
     @groups = current_user.groups.includes(:users)
     @message = Message.new
-    @messages = Message.where(group_id: params[:group_id]).order('created_at DESC').includes(:user)
+    @messages = Message.where(group_id: params[:group_id]).order('created_at ASC').includes(:user)
```
@groupを定義して、@group.messagesと書いたほうがきれいですね
また、created_at ASCも不要だと思うのですがいかがでしょうか？

robot269006 replied:
created_at ASCは以前にご指摘を受けて修正したのですが、昇順の方が投降後にすぐ最新メッセージをみれるのでその方がいいと思ってます。そのままでも問題ないでしょうか？

robot269006 replied:
@group.messagesの方はどのように書くのでしょうか？色々試してはみたのですが、どうも意図する表示ができません。。

yusukeyamane replied:
@groupはどの様に定義されましたか？？

robot269006 replied:
@groupはbefore_actionで以下のset_groupを走らせるようにしてます。
  def set_group
    @group = Group.find(params[:group_id])
  end

seoky0 replied:
>意図する表示ができない

こちらの部分をより具体的に教えていただけますか？
エラーが出るのか、ただ表示がされないのかなどをお教えていただければ幸いです。

robot269006 replied:
すみません、エラー解決しました。@group.messagesではなく、@groups.messagesになってました。。完全な凡ミスです、申し訳ありません。

robot269006 committed:
Revision (gitignore)

robot269006 committed:
Revision (indent)

robot269006 committed:
Revision (Refactor messages_controller)
